src/library_wasi.js

```diff
@@ -227,34 +228,49 @@ var WasiLibrary = {
     var num = SYSCALLS.doWritev(stream, iov, iovcnt, offset_low);
     {{{ makeSetValue('pnum', 0, 'num', 'i32') }}};
     return 0;
+#elif ASSERTIONS
+    abort('fd_pwrite called without SYSCALLS_REQUIRE_FILESYSTEM');
```

Comment:
This is more of an internal compiler error than something we want to check at runtime, I think - it means we are not linking in JS properly. Perhaps we should just have ```#error ".."``` in these places?

Reply:
No, I think all of these need to runtime errors since `-sFILESYSTEM=0` doesn't prevent the usage of the syscalls.. its just makes them error if called.   At least that is the current behaviour.

Reply:
We could just the behaviour of `-sFILESYSTEM=0` but that would be a larger change.

Reply:
That is true for `FILESYSTEM=0`, but it is not the same as `SYSCALLS_REQUIRE_FILESYSTEM`. That internal setting is set when we think that no syscalls require the filesystem, and then we set `FILESYSTEM=0` and so forth. So if `SYSCALLS_REQUIRE_FILESYSTEM` is set but we do need some non-trivial syscall, we've taken a wrong path internally in the compiler (somewhere in `emscripten.py` iirc is that logic).

Reply:
`SYSCALLS_REQUIRE_FILESYSTEM` is also set based on the user specifying `-sFILESYSTEM=0`:

```
  if settings.FILESYSTEM == 0:                                                                    
    # without filesystem support, it doesn't matter what syscalls need
    settings.SYSCALLS_REQUIRE_FILESYSTEM = 0   
```

Currently its possible to build a program that uses `fopen` with `-sFILESYSTEM=0`.. and this will compile and run and only cause a error at runtime if `fopen` is actually called.

And because `-sFILESYSTEM=0` implies `SYSCALLS_REQUIRE_FILESYSTEM==0` we cannot currently make fopen usage in this case into a build error.

Reply:
Another way of putting it SYSCALLS_REQUIRE_FILESYSTEM==0 does not currently ever produce a build error in an otherwise build-able program

Reply:
Another way of putting: SYSCALLS_REQUIRE_FILESYSTEM==0 seems to currently translate to "include stdout basics only, and error at runtime on other filesystem operations".

Reply:
I see, good points.